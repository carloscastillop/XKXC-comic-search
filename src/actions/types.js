export const FETCH_COMIC = 'FETCH_COMIC';
export const FETCH_ERROR = 'FETCH_ERROR';
