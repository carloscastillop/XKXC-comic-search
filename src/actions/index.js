import {LATEST_COMIC, FETCH_COMIC, FETCH_ERROR} from './types';
import axios from 'axios';

const apiUrl = 'https://xkcd.now.sh/?comic='; //latest or 1000

export const fetchComic = (comic) => {
    return {
        type: FETCH_COMIC,
        comic
    };
};

export const latestComic = (comic) => {
    return {
        type: LATEST_COMIC,
        comic
    };
};

export const fetchError = (error) => {
    return {
        type: FETCH_ERROR,
        error
    };
};

export const fetchComics = (num = 'latest') => {
    return (dispatch) => {
        dispatch(fetchComic(null));
        dispatch(fetchError(false));
        return axios.get(apiUrl+num, {
            crossDomain: true,
            dataType: 'json',
            logLevel: 'debug',
            changeOrigin: true
        })
            .then(response => {
                dispatch(fetchComic(response.data));
                if(num === 'latest'){
                    dispatch(latestComic(response.data));
                }
            })
            .catch(error => {
                dispatch(fetchError(true));
                throw(error);
            });
    };
};
