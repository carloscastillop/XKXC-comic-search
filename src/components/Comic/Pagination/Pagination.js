import React from 'react';
import style from './Pagination.module.scss';
import {Link} from 'react-router-dom';
import PropTypes from 'prop-types';

const Pagination = ({num, latest}) => {
    if (!num) {
        return null;
    }
    const maxNum = parseInt(latest);

    let previousBtn = (
        <li className={`${style['page-item']} ${(num - 1 > 0) ? '' : style['disabled']+' '+style['pageDisabled'] }`}>
            <Link
                className={style['page-link']}
                to={`/${num - 1}`}
                tabIndex="-1"
                aria-disabled={(num - 1 > 0) ? '' : true}
            >
                Previous
            </Link>
        </li>
    );

    let nextBtn = (
        <li className={`${style['page-item']} ${(num + 1 > maxNum) ? style['disabled']+' '+style['pageDisabled'] : ''}`}>
            <Link
                className={style['page-link']}
                to={`/${num + 1}`}
                aria-disabled={(num + 1 >= maxNum) ? true : ''}
            >
                Next
            </Link>
        </li>
    );

    return (
        <div data-testid={'pagination'} className={`${style.comicPagination} ${style['mt-5']}`}>
            <nav aria-label="Comics page navigation">
                <ul className={`${style['pagination']} ${style['justify-content-center']}`}>
                    {previousBtn}
                    {nextBtn}
                </ul>
            </nav>
        </div>
    );
};

export default Pagination;

Pagination.propTypes = {
    num: PropTypes.number,
    latest: PropTypes.number
};


