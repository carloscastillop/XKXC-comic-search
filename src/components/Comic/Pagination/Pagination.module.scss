@import '../../../styles/global.module';

.comicPagination {
  position: fixed;
  bottom: 0;
  left: 0;
  width: 100%;
  padding: $spacer;
  background-color: rgba(0, 0, 0, 0.8);

  nav {
    ul {
      margin: 0;

      a {
        background-color: $secondary;
        border-color: rgba(255, 255, 255, 0.2);
        color: #fff;

        &:hover {
          background-color: #222;
          color: #fff;
        }
      }

      .pageDisabled a, .pageDisabled a:hover {
        background-color: $secondary !important;
        border-color: rgba(255, 255, 255, 0.2) !important;
        color: rgba(255, 255, 255, 0.1) !important;
        cursor: not-allowed !important;
      }
    }
  }
}