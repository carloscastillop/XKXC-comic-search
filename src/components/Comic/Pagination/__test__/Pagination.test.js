import React from 'react';
import {render} from '@testing-library/react';
import renderer from 'react-test-renderer';
import Pagination from '../Pagination';
import {Provider} from 'react-redux';
import {applyMiddleware, createStore} from 'redux';
import rootReducer from '../../../../reducers';
import thunk from 'redux-thunk';
import {fetchComic} from "../../../../actions";
import {BrowserRouter} from "react-router-dom";

const comic = {
    month: '6',
    num: 2322,
    link: '',
    year: '2020',
    news: '',
    safe_title: 'ISO Paper Size Golden Spiral',
    transcript: '',
    alt: 'The ISO 216 standard ratio is cos(45Â°), but American letter paper is 8.5x11 because it uses radians, and 11/8.5 = pi/4.',
    img: 'https://imgs.xkcd.com/comics/iso_paper_size_golden_spiral.png',
    title: 'ISO Paper Size Golden Spiral',
    day: '19'
};

const store = createStore(
    rootReducer,
    applyMiddleware(thunk)
);

test('Pagination container', () => {
    store.dispatch(fetchComic(comic));
    const {getByTestId} = render(
        <BrowserRouter>
            <Pagination num={1} latest={10}/>
        </BrowserRouter>
    );

    const linkElement = getByTestId('pagination');
    expect(linkElement).toBeInTheDocument();
});

test('Pagination snapshop', () => {
    const tree = renderer
        .create(
            <BrowserRouter>
                <Pagination num={1} latest={10}/>
            </BrowserRouter>
        )
        .toJSON();
    expect(tree).toMatchSnapshot();
});