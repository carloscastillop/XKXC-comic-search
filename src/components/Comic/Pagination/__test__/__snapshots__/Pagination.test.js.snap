// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination snapshop 1`] = `
<div
  className="comicPagination mt-5"
  data-testid="pagination"
>
  <nav
    aria-label="Comics page navigation"
  >
    <ul
      className="pagination justify-content-center"
    >
      <li
        className="page-item disabled pageDisabled"
      >
        <a
          aria-disabled={true}
          className="page-link"
          href="/0"
          onClick={[Function]}
          tabIndex="-1"
        >
          Previous
        </a>
      </li>
      <li
        className="page-item "
      >
        <a
          aria-disabled=""
          className="page-link"
          href="/2"
          onClick={[Function]}
        >
          Next
        </a>
      </li>
    </ul>
  </nav>
</div>
`;
